src/box/txn_limbo.c

```diff
@@ -649,10 +662,66 @@ txn_limbo_read_demote(struct txn_limbo *limbo, int64_t lsn)
 	return txn_limbo_read_promote(limbo, REPLICA_ID_NIL, lsn);
 }
 
+/**
+ * Check that some synchronous transactions have gathered quorum and
+ * write a confirmation entry of the last confirmed transaction.
+ */
+void
+txn_limbo_confirm(struct txn_limbo *limbo)
+{
+	if (limbo->is_in_rollback)
+		return;
+	struct txn_limbo_entry *e;
+	if (!limbo->confirmed_entry_is_valid) {
+		/* Try to find the next valid confirmed_entry value. */
+		e = rlist_next_entry(limbo->confirmed_entry, in_queue);
+		if (rlist_entry_is_head(e, limbo->queue, in_queue))
+			return;
+		for (; !rlist_entry_is_head(e, limbo->queue, in_queue);
+		       e = rlist_next_entry(e, in_queue))
+			if (txn_has_flag(e->txn, TXN_WAIT_ACK))
+				break;
+		limbo->confirmed_entry = rlist_prev_entry(e, in_queue);
+		/*
+		 * There is no next transaction or the next transaction has no
+		 * lsn. Therefore, confirmed_entry cannot be considered valid.
+		 */
+		if (rlist_entry_is_head(e, limbo->queue, in_queue) || e->lsn == -1)
```

Comment:
How can `e` be the head, if on the line 678 you did an early exit with that condition, and since then `e` was only moving forwards (towards the tail)? I added `assert(!rlist_entry_is_head(e, limbo->queue, in_queue));` and tests passed. Does it mean the `if` can be simplified?

Reply:
As I see `rlist`'s tail - his head.